File: code/c/05-linear_search.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/**
 * Function to perform linear search on an array.
 *
 * @param arr The array to search in.
 * @param n The size of the array.
 * @param x The target value to search for.
 * @return The index of the target if found, otherwise -1.
 */
int linear_search(int arr[], int n, int x) {
  for (int i = 0; i < n; i++) {
    if (arr[i] == x) return i;  // Element found at position i
  }
  return -1;  // Element not found
}

/**
 * Main function to demonstrate linear search and measure execution time.
 */
int main() {
  int N_values[] = {1000000, 2000000, 4000000, 8000000, 16000000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    int *arr = (int *)malloc(N * sizeof(int));

    // Fill the array with sequential elements
    for (int i = 0; i < N; i++) {
      arr[i] = i;
    }

    // Set the target value to be at the end of the array
    // (worst case for linear search)
    int target = N - 1;

    // Measure the start time
    clock_t start_time = clock();

    // Perform linear search
    int result = linear_search(arr, N, target);

    // Measure the end time
    clock_t end_time = clock();

    // Calculate the elapsed time in seconds
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Print the result and time
    printf("Linear Search with N = %d\n", N);
    if (result != -1)
      printf("Target found at index %d\n", result);
    else
      printf("Target not found\n");
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free the allocated memory
    free(arr);
  }

  return 0;
}

File: code/zig/05-linear_search.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const allocator = std.heap.page_allocator;

/// Function to perform linear search on an array.
///
/// # Parameters
///
/// - `arr`: The array to search in.
/// - `x`: The target value to search for.
///
/// # Returns
///
/// The index of the target if found, otherwise `null`.
fn linearSearch(arr: []const usize, x: usize) ?usize {
    for (0.., arr) |index, item| {
        if (item == x) return index; // Element found at position index
    }
    return null; // Element not found
}

/// Main function to demonstrate linear search and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 1_000_000, 2_000_000, 4_000_000, 8_000_000, 16_000_000 };

    for (N_values) |N| {
        // Allocate memory for the array
        var arr = try allocator.alloc(usize, N);
        defer allocator.free(arr);

        // Fill the array with sequential elements
        for (0..N) |i| {
            arr[i] = i;
        }

        // Set the target value to be at the end of the array
        // (worst case for linear search)
        const target: usize = N - 1;

        // Measure the start time
        const start = try Instant.now();

        // Perform linear search
        const result = linearSearch(arr[0..], target);

        // Measure the end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Print the result and time
        print("Linear Search with N = {d}\n", .{N});
        if (result) |found| {
            print("Target found at index {d}\n", .{found});
        } else {
            print("Target not found\n", .{});
        }
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}
